File: C++/11-Container-With-Most-Water.cpp
#include<vector>
#include<algorithm>
using namespace std;

// 使用前后两个指针l和r，遍历每一种宽度得到的最大矩形区域面积。
// 我们只会将较短的一边往里移动，因为只有这样才可能得到更大面积的矩形。
class Solution {
public:
    int maxArea(vector<int>& height) {
        int max_area = 0, l = 0, r = height.size() - 1;
        while (l < r) {
            max_area = max(max_area, min(height[l], height[r]) * (r - l));
            if (height[l] < height[r]) {
                ++l;
            } else {
                --r;
            }
        }
        return max_area;
    }
};

File: Python/11-Container-With-Most-Water.py
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        l, r = 0, len(height) - 1
        res = 0
        while l < r:
            left, right = height[l], height[r]
            area = min(left, right) * (r - l)
            # update res
            res = area if area > res else res
            if left < right:
                l += 1
            else:
                r -= 1
        return res

if __name__ == '__main__':
    tmp = Solution()
    height = [1,8,6,2,5,4,8,3,7]
    res = tmp.maxArea(height)
